### DISCUSSION TITLE
Difference between "update including create" and "create+update"?

### DISCUSSION THREAD

**Discussion by TheLexoPlexx** [prisma/prisma#18468] [Q&A] [answered]
Hi there, I've got a relatively simple problem and I was looking for the most elegant solution:

The Model looks like this:

```prisma


model Zeitraum {
  id String @id @default(cuid())

  abwesenheit   Abwesenheit?
  abwesenheitId String?
}

model Abwesenheit {
  id        String          @id @default(cuid())
  summeTage Int

  zeitraum            Zeitraum @relation(fields: [zeitraumId], references: [id])
  zeitraumId          String   @unique
}
```

Because I needed "abwesenheit" to be optional on Zeitraum but required on Abwesenheit.

What I then did was to use an update-function and create a new Abwesenheit within, like this:

```ts
prisma.zeitraum.update({
  where: {
    id: req.body.zeitraum
  },
  data: {
    abwesenheit: {
      create: {
        summeTage: 0,
        typ: {
          connectOrCreate: {
            create: {
                name: req.body.reason
            },
            where: {
                name: req.body.reason
            }
          }
        }
      }
    }
  },
  include: {
    abwesenheit: true
  }
})
```

which seems to be valid syntax, however, when running, I get the following error:

```log
error - Error: Invalid `prisma.zeitraum.update()` invocation:

The change you are trying to make would violate the required relation 'AbwesenheitToZeitraum' between the `Abwesenheit` and `Zeitraum` models.
```

And with a bit of AI-Help, I ended up with this:
```ts
prisma.abwesenheit.create({
  data: {
    zeitraum: {
      connect: {
        id: req.body.zeitraum
      }
    },
    summeTage: 0,
  }
}).then((pAbwesenheit) => {

  prisma.zeitraum.update({
    where: {
      id: req.body.zeitraum
    },
    data: {
      abwesenheit: {
        connect: {
          id: pAbwesenheit.id
        }
      }
    },
    include: {
      abwesenheit: true
    }
  })
})
```
but this solution does not really satisfy me and to my understanding, both do the exact same thing. I get that both sides need to be updated db-wise but this is what I expected the "connect"-feature to do.

Why is the first and the latter version not equal?

**Comment by ludralph** [comment, marked as answer]
Hi @TheLexoPlexx 👋 

The second version is creating a new `Abwesenheit` object using the `create` method, and then connecting it to an existing `Zeitraum` object using the `connect` method. it guarantees that only one `Abwesenheit` object is associated with a given Zeitraum object, as required by the relation between the two models. However in the first version, you are using the `create` method to create a new `Abwesenheit` object and associate it with an existing `Zeitraum` object, using the `abwesenheit` field. However, this approach can cause a conflict if the `Zeitraum` object already has an associated `Abwesenheit` object.

**Reply by TheLexoPlexx** [reply]
First up, thank you @ludralph

That makes sense to me, this has lead me to the conclusion that I need a forced-update/create-overwrite or... upsert, which means I came up with the following:

```ts
prisma.zeitraum.update({
  where: {
    id: req.body.zeitraum
  },
  data: {
    abwesenheit: {
      upsert: {
        create: {
          summeTage: 0,
        },
        update: {
          summeTage: 0,
        }
      }
    }
  },
  include: {
    abwesenheit: true
  }
})
```

What is the reasoning behind having to set the values for create and update twice?

btw. this github-discussions-Page is about a million times better than some discord-Server or an additional forum.

**Reply by ludralph** [reply]
In the `upsert` method, you need to specify both the `create` and `update` objects because `upsert` is actually a combination of `create` and `update` operations.

When you call `upsert`, Prisma first checks if the data you want to update already exists. If it does, Prisma performs an update operation using the `update` object you provided. If it doesn't exist, Prisma creates a new object using the `create` object you provided.

So by setting the values for both `create` and `update` objects, you are telling Prisma what to do in both cases: if the Abwesenheit object already exists, update its summeTage field to 0, and if it doesn't exist, create a new Abwesenheit object with a summeTage field set to 0. This approach ensures that the correct data is always present in the database and avoids any potential conflicts that might arise from attempting to create or update an object that already exists. See the [docs](https://www.prisma.io/docs/concepts/components/prisma-client/crud#update-or-create-records) for upsert to learn more.

Finally, thank you for your kind words.
